### ISSUE TITLE
Mirror shows no cache entries for v0.16.2-3526-e6e21b1 but can retrieve cache entries using other versions

### ISSUE THREAD

**Issue by ndcontini** [LLNL/spack#24886]
I noticed that when using v0.16.2-3526-e6e21b1, it appears as though my mirror has no build cache entries. However, if I switch to v0.16.2 or v0.16.1, I am able to see cache entries.

### Steps to reproduce the issue

```console
$ git clone https://github.com/spack/spack
$ cd spack
$ source share/spack/setup-env.sh
$ spack mirror add mvapich2-gdr http://mvapich.cse.ohio-state.edu/download/mvapich/spack-mirror/mvapich2-gdr
$ spack buildcache list -L -v -a
==> 0 cached builds.
$ git checkout v0.16.2
$ spack buildcache list -L -v -a
==> 106 cached builds.
...
```

### Information on your system

* **Spack:** 0.16.2-3526-e6e21b1
* **Python:** 3.6.8
* **Platform:** linux-rhel6-x86_64
* **Concretizer:** original

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
This is really hard to reproduce given the information in the description. Is it still an issue with recent versions of Spack?

**Comment by ndcontini** [comment]
I figured out why this issue occurred. Essentially, between v0.16.2 and that development version of spack, there were some formatting changes to the cache index. The old format was not forwards-compatible and the new format was not backwards-compatible. This meant that you had to use a version that used the same format as the version of spack used to generate the cache. I would say the only "issue" here is that there was no error message to communicate this to me as the end user (I only discovered the reason for the missing cache entries after enabling debugging output). Otherwise, I'd say this specific issue is resolved.
